Introduce LazyLoad Backend

paarthmadan opened the pull request:
**Note: I'm using this PR to collect feedback internally. I'll close it once we're aligned and propose this upstream.**

### What's in this PR

This PR introduces a new `LazyLoad` backend following this discussion in https://github.com/ruby-i18n/i18n/issues/592. 

### What does the `LazyLoad` Backend offer?

The `Simple` backend can't infer which files belong to which locale, so it loads all files in the load path and resolves the locale by inspecting the translations that are loaded. This is a fool proof strategy, but it comes at the expense of needing to **load all** files for all locales, for any arbitrary locale.

This backend avoids the cost of loading unnecessary translation files by carefully selecting only those files which are needed for the current locale. It lazily initializes translations on a per locale basis.

### How does the `LazyLoad` Backend work?

This backend trades off the expensive cost of `I/O` with the cost of perform string matching on files in the load path. It makes assumptions about which files belong to a locale and selectively loads only these files.

### How does the `LazyLoad` Backend know which files belong to which locale?

It makes assumptions about how files are named. Clients must abide by this naming system if they decide to use this backend.

The heuristic used to bind a file to its locale can be defined as follows:
1. the filename is in the I18n load path
2. the filename ends in a supported extension (ie. .yml, .json, .po, .rb)
3. the filename starts with the locale identifier (ex: "translations/en_001.yml")

### When should someone use this backend?

Workloads that operate in the context of a single locale at a time and have many translations files for many locales. For instance, a large Rails workload would benefit from this backend.

It's designed for test environments, not environments where eager loading is preferred.

### Benchmarks: Comparing the `Simple` backend to the `LazyLoad` backend

A benchmark setup was used to compare the performance of these two backends.

**Table 1: Setup with 10 files per locale, 100 keys in each file:**

| Backend  | Work Performed                | User     | Sys      | Total    | Real     |
|----------|-------------------------------|----------|----------|----------|----------|
| Simple   | Eager load (:en)              | 0.012764 | 0.000721 | 0.013485 | 0.013503 |
| Simple   | 3 Eager loads (:en, :fr, :de) | 0.012364 | 0.000675 | 0.013039 | 0.013038 |
| LazyLoad | Eager load (:en)              | 0.004820 | 0.000330 | 0.005150 | 0.005137 |
| LazyLoad | 3 Eager loads (:en, :fr, :de) | 0.019816 | 0.000847 | 0.020663 | 0.020674 |

**Table 2: Setup with 100 files per locale, 1000 keys in each file:**

| Backend  | Work Performed                | User     | Sys      | Total    | Real     |
|----------|-------------------------------|----------|----------|----------|----------|
| Simple   | Eager load (:en)              | 1.342190 | 0.020641 | 1.362831 | 1.363569 |
| Simple   | 3 Eager loads (:en, :fr, :de) | 1.344860 | 0.018035 | 1.362895 | 1.363284 |
| LazyLoad | Eager load (:en)              | 0.478600 | 0.011205 | 0.489805 | 0.489951 |
| LazyLoad | 3 Eager loads (:en, :fr, :de) | 1.357584 | 0.026064 | 1.383648 | 1.384148 |

### Exploring the results

The `Simple` backend works for the same amount of time in the case when it needs to load translations for a single locale, **and** when it loads translations for all locales. This makes sense as the backend loads all translations irrespective of the current locale.

The `LazyLoad` backend reduces working time as it avoids loading unnecessary files. In the case when loading for a single locale, we see that the `LazyLoad` backend outperforms  `Simple`, 0.005 vs 0.013 in Table 1 and 0.4899 vs 1.363 in Table 2.

The `LazyLoad` backend performs roughly on-par with the `Simple` backend when it needs to load all translations. There is additional overhead of string matching which brings down the performance in small workloads. It's negligible in any significant workloads compared to the time spent in `I/O`.

### Remarks

This backend is designed to bring performance improvements to workloads with a large volume of locales, translation files, and translation keys.

Performance isn't guaranteed for all applications, which is why the backend is designed to be opt-in. 

At Shopify, we've patched `ruby-i18n` locally to implement a similar strategy. We've observed close to **10x speed ups locally in specific tests and roughly 20% speeds across the suite.**

paarthmadan committed:
Sort imports

paarthmadan committed:
Introduce LazyLoad backend

paarthmadan committed:
Benchmark: Compare LazyLoad vs Simple backend

paarthmadan commented on test/benchmark/benchmark_lazyload_test.rb:
```diff
@@ -0,0 +1,61 @@
+require 'test_helper'
```
This is a temporary test that's used to setup the benchmarks. It'll be presented upstream, but I don't intend on merging this.

paarthmadan commented:
Please review @Shopify/rails

cc: @adrianna-chang-shopify, @shioyama

adrianna-chang-shopify commented on test/backend/lazy_load_test.rb:
```diff
@@ -0,0 +1,107 @@
+
+  private
+
+  def with_temp_file_in_load_path(name, tmpdir, file_contents, &block)
```
As discussed IRL, I think we can drop the block parameter here and `yield` instead of calling the block, since we expect this to always be used with a block 😄

adrianna-chang-shopify commented on lib/i18n/backend/lazy_load.rb:
```diff
@@ -0,0 +1,92 @@
+        end
+      end
+
+      SUPPORTED_EXTENSIONS = [".yml", ".yaml", ".po", ".json", ".rb"].freeze
```
We discussed potentially extracting this constant into [`I18N::Base`](https://github.com/ruby-i18n/i18n/blob/master/lib/i18n/backend/base.rb) and checking for the presence of the extension in the list rather than relying on the [`respond_to` pattern](https://github.com/ruby-i18n/i18n/blob/master/lib/i18n/backend/base.rb#L221-L229) to load translation files.

adrianna-chang-shopify commented on test/backend/lazy_load_test.rb:
```diff
@@ -0,0 +1,107 @@
+    invalid_files.each do |file|
+      with_temp_file_in_load_path(file[:filename], file[:dir], file_contents) do
+        I18n.with_locale(:en) { @backend.eager_load! }
+
```
✂️ 
```suggestion
```

adrianna-chang-shopify commented on test/backend/lazy_load_test.rb:
```diff
@@ -0,0 +1,107 @@
+    valid_files.each do |file|
+      with_temp_file_in_load_path(file[:filename], file[:dir], file_contents) do
+        I18n.with_locale(:en) { @backend.eager_load! }
+
```
✂️ 
```suggestion
```

adrianna-chang-shopify commented on test/backend/lazy_load_test.rb:
```diff
@@ -0,0 +1,107 @@
+    ]
+
+    invalid_files.each do |file|
+      with_temp_file_in_load_path(file[:filename], file[:dir], file_contents) do
```
Nit: Can we give this a better name? How about `with_translation_file_in_load_path` or `with_translation_file_loaded`?

adrianna-chang-shopify reviewed (commented):
This is great work, Paarth! 👏  🚀 

As discussed IRL: `#available_locales` doesn't work right now, because it will only look at loaded locales, which is likely to be incomplete. I think we'll need to do something similar to what @shioyama did in Core and use our selection heuristics to grab available locales from the load path.

A couple of points we talked about IRL that are not blockers for this PR, but that we might want to think about in terms of delivering this as a feature upstream. I've writing them down here in case anyone else has thoughts / feedback on them, and to remind myself what was discussed 😄 
1) We might want to ensure that this can work out of the box with Rails. There are default translations in Rails that don't conform to the path-matching regex we've specified. Maybe this means additional support in https://github.com/svenfuchs/rails-i18n to ensure that these translations are always loaded by default, similar to what we did in Core.
2) Should we have some sort of sanity check that apps can use to verify their translations are set up correctly? Might be nicer than failing silently if a translation has an unusual file path and doesn't get picked up properly. Possibly a rake task?

I don't have sufficient context on the failing JRuby tests -- maybe someone else from the team can step in there.

~~One more question: are we intending that folks use this in production? Obviously the use case we have in Core is strictly for our test environment.~~ Prod is eager loaded so this is a no-go 🤦‍♀️ 

I think our next step is to try to adopt this in Core, and make sure all the tests continue to pass as expected.

shioyama commented on lib/i18n/backend/lazy_load.rb:
```diff
@@ -0,0 +1,92 @@
+      end
+
+      def basename_starts_with_locale?(path)
+        File.basename(path).start_with?(I18n.locale.to_s)
```
A file like `pt-BR.yml` will be loaded if the locale is `pt`, but also (again) if the locale is switched to `pt-BR`, since both match `"pt-BR.yml".start_with?(...)`. This will lead to repeat-loading which is inefficient and potentially can lead to duplicate warnings and other bugs.

It feels like we need a separator here which is not also used as part of locales themselves (i.e. something non-alphanumerical and also not `-`).

Assuming that separator is `SEPARATOR` (say `_`), then a way to do this would be:

```suggestion
        File.basename(path, ".*").split(SEPARATOR, 2)[0] == I18n.locale.to_s
```

Then `basename_starts_with_locale?("foo/pt-BR.yml")` would be true for `pt-BR` but not for `pt`. Likewise if `SEPARATOR` was `_`, then `basename_starts_with_locale?("foo/pt-BR_translation.yml")` would also be true for `pt-BR` (but not `pt`).

In the end, without some separator that is not `-`, we'll be double-loading locale files. Having a clear separator also allows us to reliably extract locales from filenames, rather than just matching against them.

shioyama commented on lib/i18n/backend/lazy_load.rb:
```diff
@@ -0,0 +1,92 @@
+      end
+
+      def path_contains_locale_identifier?(path)
+        Pathname.new(path).each_filename.any? { |component| component == I18n.locale.to_s }
```
Playing the devil's advocate here, a file structure like `en/fr/de/` will lead to double-loading of the same files for different locales. I think that's fine; folks using this backend should know better than to use that kind of directory structure. But just to point out that the potential to double-load translations is there.

shioyama commented on lib/i18n/backend/lazy_load.rb:
```diff
@@ -0,0 +1,92 @@
+    # Valid files that will be selected by this backend:
+    #
+    # "files/locales/en-translation.yml" (Selected for locale "en")
+    # "files/locales/fr/translation.po"  (Selected for locale "fr")
```
If we end up implementing `available_locales` (and I think we will need to), I wonder how we will select locales from load paths. 🤔  Currently in core we just ignore the latter pattern and look for the former pattern, narrowed to locales which match `%r{/([A-z]{2,3}(\-[A-z]{2,3})?)\.yml\z}`. This is a subset that happens to work reasonably enough for our needs there but would not work in general.

Without being able to make assumptions, we'd have to have a clearer demarcation of the locale in the name in the `en-translation.yml` format (which I mentioned in another comment). Otherwise it's impossible to extract locales from filenames in general (i.e. `pt-BR.yml` would be interpreted as including both `pt` and `pt-BR` locales).

In the latter format, we could take all components of each filename and filter those which are a valid locale, maybe with some config to exclude invalid locales?

Either way I think it's important we consider `available_locales` beforehand since once we settle on a format convention it will be hard to go back if it ends up not working well for this use case.

paarthmadan replied:
These are great points, and for the most part, all of your comments have helped shaped the new solutions and the assumptions we make 👍

shioyama reviewed (commented):
Looks great! 👍 Just some initial comments, mainly about the format conventions for locale filenames.

shioyama commented on lib/i18n/backend/lazy_load.rb:
```diff
@@ -0,0 +1,92 @@
+        I18n.load_path.flatten.select do |path|
+          (basename_starts_with_locale?(path) || path_contains_locale_identifier?(path)) &&
+          supported_extension?(path)
+        end
```
I wonder if it would be worthwhile to track which locale files have already been loaded, to prevent them from being loaded twice. Both as a safeguard and as an edge-case perf optimization.

e.g. something like this:

```diff
       def reload!
         @initialized_locales = nil
         @translations = nil
+        @loadable_paths = nil
       end
 
+      def load_translations(*filenames)
+        @loadable_paths = loadable_paths - filenames.flatten
+        super
+      end
+
       protected
 
       # Load translations from files that belong to the current locale.
@@ -64,11 +70,15 @@ module I18n
         @initialized_locales ||= Hash.new(false)
       end
 
+      def loadable_paths
+        @loadable_paths ||= I18n.load_path.flatten
+      end
+
       # Select all files from I18n load path that belong to current locale.
       # These files must start with the locale identifier (ie. "en", "fr").
       # or contain a path component equal to the locale identifier (ie. /locales/en/translation.yml)
       def filenames_for_current_locale
-        I18n.load_path.flatten.select do |path|
+        loadable_paths.select do |path|
           (basename_starts_with_locale?(path) || path_contains_locale_identifier?(path)) &&
           supported_extension?(path)
         end
```

So basically when you load translations, you remove those paths from `loadable_paths` and even if there is some overlap with another locale, you don't load them again unless the backend has been reloaded.

I don't want to add complexity if unnecessary so this is just an idea.

adrianna-chang-shopify replied:
This seems like a pretty reasonable thing to add IMO! Definitely helps guard against people doing weird things with their directory structures 😄

shioyama replied:
Yeah I think having this kind of safeguard would allow us to be more loose with formats, so e.g. my earlier point about `pt.yml` vs `pt-BR.yml` wouldn't be such a big deal if this were in place: `pt` would load both, but if you switched to `pt-BR` it would not load that locale's translations again.

paarthmadan replied:
I've removed the latter format (ie. locale in path components), and enforced the presence of a demarcation ("_") if text proceeds the locale.

This simplifies the assumptions, ie.
1. We can infer `#available_locales` concretely
2. We don't need to safeguard against multiple loads (although this solution was 👍 )

Thanks for your help.

casperisfine commented:
> When should someone use this backend?

I believe you should make it clear in that section that this backend is for development and test environment, and shouldn't be used in production environments.

casperisfine commented on lib/i18n/backend/lazy_load.rb:
```diff
@@ -0,0 +1,92 @@
+      SUPPORTED_EXTENSIONS = [".yml", ".yaml", ".po", ".json", ".rb"].freeze
+
+      def supported_extension?(path)
+        SUPPORTED_EXTENSIONS.include?(File.extname(path))
```
Not sure how much of a hotpath this is, but my hunch says this could be faster:

```suggestion
        path.end_with?(*SUPPORTED_EXTENSIONS)
```

Did a quick bench:

```ruby
# frozen_string_literal: true

require 'benchmark/ips'

SUPPORTED_EXTENSIONS = [".yml", ".yaml", ".po", ".json", ".rb"].freeze

Benchmark.ips do |x|
  x.report("end_with?") { "/Users/byroot/src/github.com/Shopify/shopify/config/database.yml".end_with?(*SUPPORTED_EXTENSIONS) }
  x.report("include? ext") { SUPPORTED_EXTENSIONS.include?(File.extname("/Users/byroot/src/github.com/Shopify/shopify/config/database.yml")) }
  x.compare!
end
```

```
Warming up --------------------------------------
           end_with?     1.128M i/100ms
        include? ext   181.313k i/100ms
Calculating -------------------------------------
           end_with?     11.262M (± 0.2%) i/s -     56.396M in   5.007837s
        include? ext      1.811M (± 0.3%) i/s -      9.066M in   5.006622s

Comparison:
           end_with?: 11261609.5 i/s
        include? ext:  1810746.8 i/s - 6.22x  (± 0.00) slower
```

paarthmadan replied:
Thanks for the benchmark, changed 👍

casperisfine commented on test/backend/lazy_load_test.rb:
```diff
@@ -0,0 +1,107 @@
+  def setup
+    super
+    @backend  = I18n.backend = I18n::Backend::LazyLoad.new
+    I18n.load_path = [locales_dir + '/en.yml', locales_dir + '/fr.yml']
```
To avoid double `/` if `locales_dir` have a trailing slash:
```suggestion
    I18n.load_path = [File.join(locales_dir, 'en.yml'), File.join(locales_dir, 'fr.yml')]
```

casperisfine commented on test/backend/lazy_load_test.rb:
```diff
@@ -0,0 +1,107 @@
+
+    assert_nil translations
+
+    I18n.with_locale(:en) { @backend.eager_load! }
```
I'm kinda torn here.

I think this backend should either raise on `eager_load!` or either load all locales because it's not supposed to be used in eager loading contexts.

That being said our CI is eager loading (and most should be https://github.com/rails/rails/pull/43508).

I'm just very worried people might accidentally use this in production, which would be bad.

adrianna-chang-shopify replied:
I think I'd opt to raise on `eager_load!` to make it clear this is not intended to be used in production / other eager loading contexts. If CI is eager loaded the performance benefits would be negligible anyways, no? (And actually probably worse than with a `SimpleBackend`, because of the extra string comparison cost). Just as was the case with us implementing this in Core, the benefits to this backend are reaped mainly for local testing.

shioyama replied:
Another option would be a loud warning, but yeah I also wondered about `eager_load!`. Raising seems reasonable to me.

paarthmadan replied:
I agree, and thanks all for the feedback—I've altered the behaviour to now raise on `eager_load!`.

paarthmadan commented:
Thanks for the reviews, I incorporated all stylistic changes, nits, and minor feedback.

> I believe you should make it clear in that section that this backend is for development and test environment, and shouldn't be used in production environments.

I agree, I've updated documentation to make this more clear, and I've changed the behaviour of `eager_load!` to solidify that this shouldn't be used in envs where eager loading is the best practice.

paarthmadan commented:
The crux of this problem, now, seems to be **which naming format the backend assumes and how we can generate available locales from this.**

I'd like to collect your feedback on some approaches for how to proceed.

Here's a small set of criteria I've used to evaluate the approaches:
1. **The solution should be generalizable**. In other words, how many apps actually abide by this format? Could a simple Rails app immediately make use of the new backend?
2. **How difficult is it to use the solution in core**?
3. How much **overhead is introduced** with this solution?

With this criteria in mind, here are some approaches Adrianna and I discussed

#### Approach 1: Assume all files abide by `<locale>-translation.yml` format (ie. locale at the start of the file)

Advantages:
- This would make the `#available_locales` implementation straightforward
- The constraint on the naming format is small, so backwards compatibility isn't severed (if we start too large, it'll be hard to move back)

Disadvantages:

- Solution isn't necessarily generalizable. (ie. many apps use paths like `/path/en/views/file.yml`)
- Usage in core isn't trivial, because not all locale files in core abide by this rule.

#### Approach 2: Assume all files abide by `<locale>-translation.yml` format OR `/path/<locale>/translation.yml` (ie. locale at the start of the file OR identifier in the path)

Advantages:
- The constraint on the naming format is still relatively small, so backwards compatibility isn't severed

Disadvantages:

- Solution is more generalizable but as [highlighted by this comment](https://github.com/Shopify/i18n/pull/3#discussion_r788417643) supporting paths like these make it hard to extract the locales)
- `#available_locales` implementation is virtually impossible without loading all translations or making hasty generalizations

#### Approach 3: Assume all files are formatted in a specified way and introduce configuration for "path to locale" resolution

This is the same solution as the two above, but now we create an interface that allows clients to specify how paths are related to their locale.

**Imagine a proc in a configuration file such as**, or something similar:

```ruby
# Example: /config/initializers/i18n.rb
I18n.lazy_load.path_resolve = ->(path) do
  extract_locale(path) # Example Implementation 1
  LookupTable[path]    # Example implementation 2
  # ... etc
end
```

Advantages:
- This helps generalization, we give the client the interface to easily work around edge cases such as files that don't abide by the naming convention.
- Enables usage in core

Disadvantages:
- Onboarding is very involved. Have to write custom logic per app to configure load path selection heuristic.
- `#available_locales` will need to be left to client, which will still be tricky.

#### Approach 4: Generate / Dump mapping between locale file and locales (use this instead of string matching)

This approach varies the most from what's already presented.

At a high-level:
1. Generate static mapping by loading all translations and relating a locale to the files that it came from.
2. This dump can be enforced in the CI build, the mapping will need to be checked into version control
3. Locally, the `LazyLoad` backend will load this mapping into memory and consult the lookup table to deduce which files need to be loaded. Instead of searching through load path, we consult the table. `#available_locales` are simply the keys of the table.
4. Locally, when new translation files are introduced or a translation files change, a re-dump will be required. I believe this can handled more intelligently (ie. If a file is modified, load it regardless of the locale)

Note: The `Simple` backend loads all translations to determine available locales anyways. We would have to pay this cost once, preferably as part of the build process. If we can define a way to resolve this mapping when local changes occur, then this solution seems promising (ie. using a digest for modified files, or always including translations that are dirty in the context of the git branch)

Advantages:
- The most generalizable. It doesn't define a specific format, or require a specific hierarchy
- Doesn't require client side configuration
- Easily works with Core

Disadvantages:
- Produces an artifact that must be persisted
- Mapping will need to be updated locally (or local changes will need to be handled differently). If this can't be done without reloading all translations locally, then it defeats the purpose.

The list certainly isn't exclusive, nor are any of the solutions well-defined. I wanted to collect feedback early to see if anyone had any strong opinions or had some ideas they'd like to share.

cc: @shioyama, @casperisfine, @adrianna-chang-shopify 

Ultimately, this problem is tricky because we don't know which files belong to a locale without loading them. The complexity rises in trying to do this without _actually_ doing it 😄

casperisfine commented on lib/i18n/backend/lazy_load.rb:
```diff
@@ -0,0 +1,100 @@
+        @translations = nil
+      end
+
+      def eager_load!
+        raise UnsupportedMethod.new(__method__, self.class)
+      end
```
So second though on this in the context of our CI (and most people CI now that Rails has eagerload by default on CI).

It might make sense to eager_load the "default locale" (typically `:en`) as it's what the vast majority of tests will end up using.

It's however tricky to reconcile the safety of preventing this from being used in production, with this extra feature. So I'm not suggesting to act on this, just braindumping a concern for which I have no solution yet.

casperisfine replied:
Actually, on CI it might make sense to eager load all locales regardless of wether they are used as it at the very least test that they don't contain syntax errors and such. So ignore the above.

adrianna-chang-shopify replied:
I wonder if we should encourage folks to use the same backend for both production and CI, like we do in Core? That way we ensure CI loads everything the same way production would, and if they have their own extensions / are using some of the pluggable I18n features those will be tested in CI.

I'm also wondering how we'll expect users to configure this. Should it just be a setting on the gem that can be opted into (ie. `I18n.lazy_load_in_test = true`), or are we expecting users to prepend the `LazyLoad` module in an initializer in their application?

casperisfine replied:
Yeah, in the context of the main question here, which is how to map a file name to a locale, and how to answer it is correct, I think it would make sense for this backend to actually have two use mode.

One lazy like today, and one eager loading mode that behave a bit like the regular one, but also make sure the assumptions about the filename to locale mapping was correct.

It would need to be renamed though, because it it wouldn't always be lazy.

adrianna-chang-shopify replied:
Gotcha. Yeah that makes sense to me -- simpler than requiring them to configure two separate backends 😄

paarthmadan replied:
Do you suggest to run a single backend in prod, CI, and local tests, but configure different modes in each environment (ie. prod => eager_load, ci => eager_load, test => lazy_load).

As I understand it, we'd use this new backend, call it `LazyLoadable` in CI and test, but in prod we continue to use the `Simple` backend.

Here's a little flow chart image to consolidate how I've viewed the ideas presented in this review:
![LazyLoad Flow Chart](https://user-images.githubusercontent.com/8986140/151243025-c1ced863-02e2-45fc-a4b9-3e824f07e14d.png)

casperisfine replied:
> Do you suggest to run a single backend in prod, CI, and local tests, but configure different modes in each environment (ie. prod => eager_load, ci => eager_load, test => lazy_load).

Yes. The idea being that when it's configured in "eager" mode, it can use the same regexp it used to assume the locales, to verify that its assumptions were correct. And if they aren't it can easily raise.

That also makes for a much simpler configuration. You always use the same backend, you just flip a switch based on the environment.

casperisfine commented:
> which naming format the backend assumes and how we can generate available locales from this.

While it's the most accurate, I don't think the mapping solution has good ergonomics. 

I think we should either enforce a naming convention and / or have a list of regexp or other type of callbacks to extract a locale from a path. This means that we should have some kind of strict mode if the assumptions end up wrong, and they should be tested via loading all ~tests~ translations.

adrianna-chang-shopify commented:
I'd lean towards approach 3. As Jean said, option 4 guarantees accuracy but is a bit tedious and involved for users to set up. I think that _especially_ because this is a test environment optimization, it should be as simple as possible for users to set up  / opt into. Going with 3 allows us to start very simple and assume that 95% of apps can comply with those expectations out of the box (I think we should start with the assumption that "all files abide by `<locale>-translation.yml` format"). We can then figure out what an ideal API looks like as we extend this to Core and figure out how to make it work there.

>Disadvantages:
>
> * Onboarding is very involved. Have to write custom logic per app to configure load path selection heuristic.
> * #available_locales will need to be left to client, which will still be tricky.

I'd actually argue that this is not so involved. For most applications, going with the default format should be enough. We can be pretty explicit about we we expect from the "path resolve"  feature -- as Jean suggested, it could be as simple as providing a regex / list of regexes that extracts the locale from the filepath. Users wouldn't necessarily need to write `#available_locales` themselves -- we should just be able to go through the files in the load path and extract the locales that meet the patterns. The biggest thing to watch out for here is matching locales that aren't really locales, but maybe we could offer a script / CLI command that "sanity checks" the app's translation setup and ensures that the list of locales provided by `#available_locales` is equivalent to the set of locale keys generated by reading all of the translation files. 

TL;DR - I think we can keep things simple for now and be rigid about the naming configuration we expect, and figure out a way to offer configuration as a follow-up.

paarthmadan commented:
After syncing with Jean and Adrianna, an approach to get a first iteration and possibly integrate with Core in the future:

1. Use simplest locale format, with either a) demarcation or b) tracking loaded_paths
2. Implement `available_locales`
3. Implement varied behaviour with configured mode (ie. `eager` vs. `lazy`)
4. Propose PR upstream to `ruby-i18n`

If the PR is well-received, the plan for Core / Rails can be as simple as:
1. Moving all locale files to abide by new format by renaming / splitting files. This can be done with a script that introspects the load path, checks if the file abides by the convention, and if not, renames and splits the file appropriately.

The amount of work remaining to propose this upstream as defined above is limited. The calendar time for reviews may be longer, but if the PR does make it upstream, the Core work can also be done at any time in the future. It should take a few days at most, mostly scripting and testing.

paarthmadan committed:
Support multiple modes, implement #available_locales, use simple format

paarthmadan committed:
Rename LazyLoad to LazyLoadable backend

casperisfine commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,155 @@
+    #
+    class LocaleExtractor
+      class << self
+        def locale_from_path(path)
+          ext = File.extname(path)
+          name = File.basename(path).chomp(ext)
+
+          locale = name.split("_").first
+          locale.to_sym unless locale.nil?
+        end
```
```suggestion
        def locale_from_path(path)
          name = File.basename(path, ".*")
          locale = name.split("_").first
          locale.to_sym unless locale.nil?
        end
```

```ruby
>> File.basename("foo.bar.txt", ".*")
=> "foo.bar"
```

casperisfine commented on lib/i18n/exceptions.rb:
```diff
@@ -110,4 +110,21 @@ def initialize(type, filename)
+    attr_reader :invalid_filenames
+
+    def initialize(invalid_filenames)
+      super "Locales cannot be extracted from the following paths: #{invalid_filenames}"
```
Might be worth adding an explanation or quick example of what the naming constraint is here.

paarthmadan replied:
👍 Made the error message more helpful

casperisfine commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,155 @@
+        else
+          super
+          filenames_named_incorrectly = I18n.load_path.reject { |path| file_named_correctly?(path) }
+          raise InvalidFilenames.new(filenames_named_incorrectly) unless filenames_named_incorrectly.empty?
```
We could raise in both lazy and eager loading. Might as well warn users that there's a problem as soon as possible.

casperisfine commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,155 @@
+      # Load translations from files that belong to the current locale.
+      def init_translations
+        if lazy_load?
+          load_translations(filenames_for_current_locale)
```
I think we should double check here that the files only contained translation for the locale we expected, e.g.:

```yaml
# fr.yml
fr:
  dog: Chien
en:
  dog: Dog
```

Should blow up.

paarthmadan replied:
Good call 👍  Added a check to ensure the recently loaded translations only contain the single locale key equal to the current path.

casperisfine reviewed (commented):
Some minor stuff, but looks like it's on the right track, and it really isn't that much code.

paarthmadan committed:
Accept block in #load_translations, yield translations

paarthmadan committed:
Raise invalid files in test, ensure file only loads for one locale

adrianna-chang-shopify commented on lib/i18n/backend/base.rb:
```diff
@@ -11,9 +11,12 @@ module Base
-        filenames.flatten.each { |filename| load_file(filename) }
+        filenames.flatten.each do |filename|
+          loaded_translations = load_file(filename)
+          block.call(filename, loaded_translations) if block_given?
```
It might be nicer to use an implicit block here (and should be faster, because Ruby won't have to create an object):
```suggestion
          yield filename, loaded_translations if block_given?
```

adrianna-chang-shopify commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,160 @@
+    #
+    # The backend has two working modes: lazy_load and eager_load.
+    #
+    # We recommend enabling this to true in test environments only.
```
Should we be more opinionated here? Something like
```suggestion
    # Note: This backend should only be enabled in test environments!
```

adrianna-chang-shopify commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,160 @@
+    #
+    #   I18n.backend = I18n::Backend::LazyLoadable.new(lazy_load: true)
+    #
+    #   # In other environments, such as Prod and CI
```
Nit 😅 
```suggestion
    #   # In other environments, such as production and CI
```

adrianna-chang-shopify commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,160 @@
+
+      # Eager loading is not supported in the lazy context.
+      def eager_load!
+        if lazy_load?
+          raise UnsupportedMethod.new(__method__, self.class)
+        else
+          super
+        end
```
We can simplify this a bit since we're going to short-circuit if we raise
```suggestion  
        raise UnsupportedMethod.new(__method__, self.class) if lazy_load?
        super
```

paarthmadan replied:
While normally I prefer the one-line conditionals, in this case because most of the methods have conditional behaviour on `lazy_load?` I wanted to unify how each of the methods look. I think it helps clearly separate the two modes and how the method is expected to behave in each case.

I don't have a strong opinion there, so I'm happy to still make the change if you feel it still makes sense 👍

adrianna-chang-shopify replied:
That's fair! Thanks for sharing your thoughts and pushing back ❤️

adrianna-chang-shopify commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,160 @@
+        @lazy_load
+      end
+
+      SUPPORTED_EXTENSIONS = [".yml", ".yaml", ".po", ".json", ".rb"].freeze
```
We might want to use `private_constant` here if we actually care about keeping this constant private

paarthmadan replied:
Good point, I've added for now. I think we can work on fixing upstream's implementation of how they solve supported extensions in a different PR. I'd like to keep this one well contained.

adrianna-chang-shopify commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,160 @@
+        invalid_files = []
+
+        if lazy_load?
+          load_translations(filenames_for_current_locale) { |file, translations| invalid_files << file unless file_named_correctly?(file, translations) }
```
Nit - I'd prefer the `do..end` syntax here since it's a "block with a side-effect" (it's generally personal preference, but there's [old blog post](https://archive.is/96ko8) about using `{}` for blocks that return values, and `do..end` for blocks that have side effects, and I think it's a common practice) 😄 
```suggestion
          load_translations(filenames_for_current_locale) do |file, translations|
            invalid_files << file unless file_named_correctly?(file, translations)
          end
```

paarthmadan replied:
Good callout, added 👍

adrianna-chang-shopify commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,160 @@
+          load_translations(filenames_for_current_locale) { |file, translations| invalid_files << file unless file_named_correctly?(file, translations) }
+          initialized_locales[I18n.locale] = true
+        else
+          load_translations { |file, translations| invalid_files << file unless file_named_correctly?(file, translations) }
```
Let's use `do..end` here too!

adrianna-chang-shopify commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,160 @@
+      # Select all files from I18n load path that belong to current locale.
+      # These files must start with the locale identifier (ie. "en", "pt-BR"),
+      # followed by an "_" demarcation to separate proceeding text.
+      def filenames_for_current_locale
```
Can we give this private visibility?

adrianna-chang-shopify commented on lib/i18n/tests/basics.rb:
```diff
@@ -5,12 +5,11 @@ def teardown
 
-        assert I18n.available_locales.include?(:de)
-        assert I18n.available_locales.include?(:en)
+        assert_equal I18n.backend.available_locales, I18n.available_locales
```
Nit: usually we order assertions as `assert_equal(expected, actual)`
```suggestion
        assert_equal I18n.available_locales, I18n.backend.available_locales
```

adrianna-chang-shopify commented on lib/i18n/tests/basics.rb:
```diff
@@ -24,8 +23,7 @@ def teardown
         I18n.available_locales = nil
-        assert I18n.available_locales.include?(:de)
-        assert I18n.available_locales.include?(:en)
+        assert_equal I18n.backend.available_locales, I18n.available_locales
```
```suggestion
        assert_equal I18n.available_locales, I18n.backend.available_locales
```

adrianna-chang-shopify commented on lib/i18n/exceptions.rb:
```diff
@@ -110,4 +110,37 @@ def initialize(type, filename)
+    def initialize(invalid_files)
+      @invalid_files = invalid_files
+      super <<~MSG
+        The following files are named incorrectly: #{invalid_files}.
```
What do you think about listing the files individually?
```suggestion
        The following files are named incorrectly:
         #{invalid_files.join("\n")}

```

adrianna-chang-shopify reviewed (commented):
This looks great, Paarth! I really appreciate the amount of detail you've put into test cases and documentation ❤️  I have a bunch of minor comments (feel free to contest any suggestions I've made that don't make sense 😆 ), and I'd like to give it a tophat on a simple Rails app, but this looks pretty much ready to be proposed upstream IMO 🚀

casperisfine commented on lib/i18n/backend/lazy_loadable.rb:
```diff
@@ -0,0 +1,160 @@
+
+      # Checks if a filename is named in correspondence to the translations it loaded.
+      # The locale extracted from the path must be the single locale loaded in the translations.
+      def file_named_correctly?(path, translations)
+        locales = translations.keys.map(&:to_sym)
+        return false unless locales.one?
+
+        LocaleExtractor.locale_from_path(path) == locales.first
+      end
```
Rather than return `true/false` this method should raise a final error. This way we can include in the error message which assumption was broken, and in which file, making it much easier to fix.

paarthmadan replied:
I was on the fence about which way to go. I agree that knowing which assumption was broken will help, and this is something I'll add with the current approach.

The reason why I decided to raise with all the offending files at once is mainly a DX concern. If someone's onboarding using this strategy and would like to acquire a complete list of offending files, it would be handy to raise this in a single error.

My worry is that people will fight with repeated errors being raised and only able to fix them on a one-by-one basis.

Thoughts?

casperisfine replied:
You are right, collecting all the offenses and raising once is even better. But I think you can do both, it just can't be expressed by simple booleans since you have multiple types of offense (e.g. path not matching at all, and content not respecting what the path claims).

So maybe you can collect a list of error message in an array and then raise if it's not empty.

adrianna-chang-shopify replied:
We reached the same conclusion when pairing on this yesterday! 😄  I believe Paarth has submitted the changes upstream now: https://github.com/ruby-i18n/i18n/pull/612/files#diff-8e52202d93e7810223b33ab64de65635e73f04ffa8d7d38b14b5e656039d3ea1R152-R162

paarthmadan committed:
Add #lookup

paarthmadan commented:
Closing in favour of https://github.com/ruby-i18n/i18n/pull/612
